### PR TITLE
Optimize sum op

### PR DESCRIPTION
**Pull request by jacquesqiao**
Currently for sparse gradient merge, we just concat the selected rows together, if we have many gradients, then the concated gradient will be very large, this will make the send/recv and optimization slow. We can just merge the gradient with the same id, so the gradient size will be small and the computation will be fast.

In one distribute test case, it will make the training much faster.

image

### CODE REVIEW

File: paddle/fluid/CMakeLists.txt [hunk 1]
```diff
@@ -14,4 +14,4 @@ if(WITH_INFERENCE)
   add_subdirectory(inference)
 endif()
 
-add_subdirectory(train)
+#add_subdirectory(train)
```
Comment (typhoonzero) [inline paddle/fluid/CMakeLists.txt]: can delete this

**Comment by typhoonzero** [comment]
Not sure whether this will change the behavior of some other models.

File: paddle/fluid/CMakeLists.txt (hunk 1 above)
Comment (jacquesqiao) [reply paddle/fluid/CMakeLists.txt]: done

File: paddle/fluid/operators/math/selected_rows_functor.cc [hunk 2]
```diff
@@ -228,8 +228,24 @@ template struct SelectedRowsAddToTensor<platform::CPUDeviceContext, int64_t>;
...
-size_t FindPos(const std::vector<int64_t>& rows, int64_t value) {
-  return std::find(rows.begin(), rows.end(), value) - rows.begin();
+template <typename DeviceContext, typename T>
+typename std::enable_if<
+    std::is_floating_point<T>::value &&
+    std::is_same<DeviceContext, platform::CPUDeviceContext>::value>::type
+elementwise_add(const DeviceContext& ctx, BlasT<DeviceContext, T>* blas,
+                size_t data_len, const T* in, T* out) {
+  blas->AXPY(data_len, 1., in, out);
+}
```
Comment (chengduoZH) [inline paddle/fluid/operators/math/selected_rows_functor.cc]: line 231~238 is not used.

File: paddle/fluid/operators/math/selected_rows_functor.cc [hunk 3]
```diff
@@ -228,8 +228,24 @@ template struct SelectedRowsAddToTensor<platform::CPUDeviceContext, int64_t>;
...
+
+template <typename DeviceContext, typename T>
+typename std::enable_if<
+    !std::is_floating_point<T>::value &&
```
Comment (chengduoZH) [inline paddle/fluid/operators/math/selected_rows_functor.cc:243]: Why do you add type check here?

Comment (jacquesqiao) [reply paddle/fluid/operators/math/selected_rows_functor.cc:243]: blas only support float and double.

File: paddle/fluid/operators/sum_op.h [hunk 4]
```diff
@@ -69,80 +69,18 @@ class SumKernel : public framework::OpKernel<T> {
...
-
-      auto get_selected_row = [&](size_t i) -> const SelectedRows & {
-        if (i == 0 && in0) {
-          return *in0.get();
-        } else {
-          return in_vars[i]->Get<SelectedRows>();
-        }
-      };
-
+      PADDLE_ENFORCE(!in_place, "SelectedRows not support inplace sum now");
```
Comment (dzhwinter) [inline paddle/fluid/operators/sum_op.h]: Why remove the inplace sum of SelectedRow?

**Review by dzhwinter — DISMISSED** [review]
LGTM

**Review by dzhwinter — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/math/selected_rows_functor.cc [hunk 5]
```diff
@@ -230,8 +229,24 @@ template struct SelectedRowsAddToTensor<platform::CPUDeviceContext, int64_t>;
...
+    std::is_same<DeviceContext, platform::CPUDeviceContext>::value>::type
+elementwise_add_to(const DeviceContext& ctx, BlasT<DeviceContext, T>* blas,
+                   size_t data_len, const T* in, T* out) {
+  for (int64_t i = 0; i < data_len; i++) {
```
Comment (velconia) [inline paddle/fluid/operators/math/selected_rows_functor.cc:247]: why don't we use ```memcpy``` here?

File: paddle/fluid/operators/math/selected_rows_functor.h [hunk 6]
```diff
@@ -83,104 +83,9 @@ struct MergeAdd {
...
-      blas.VCOPY(input_width, in_data + rows[0] * input_width, out_ptr);
-
-      for (size_t i = 1; i < rows.size(); ++i) {
-        blas.AXPY(input_width, 1., in_data + rows[i] * input_width, out_ptr);
-      }
-    }
-
-    out.set_rows(merge_rows);
-  }
+  void operator()(const DeviceContext& context,
```
Comment (velconia) [inline paddle/fluid/operators/math/selected_rows_functor.h:86]: Seems conflict with the ```SelectedRowsSumTo``` functor.

Comment (jacquesqiao) [reply paddle/fluid/operators/math/selected_rows_functor.h:86]: Merge add will merge the value with the same id, not concat.

File: paddle/fluid/operators/math/selected_rows_functor.cc (hunk 5 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/math/selected_rows_functor.cc:247]: Here is add to but not concat.

**Pull request merged**